website/docs/devtools/net-design.md

```diff
@@ -0,0 +1,486 @@
+
+On a positive note the reference implementation is in go over [here](https://github.com/ethereum/go-ethereum/tree/master/p2p/discv5), we wouldn’t have to implement it, since we can just use this version.
+
+Discovery Communication is authenticated using session keys which are established in the handshake. There are different ways to save node records, either you could utilise an in-memory db for a temporary store or you could spin off a leveldb instance which can be persisted across sessions. For Prysm it would be easier for each node to simply have an in-memory db which  stores all the node records, we can look to have a persisted store for mainnet where we can analyse trade offs to this approach. 
+
+**Listening on a Port**
+
+```go
+network, err := discv5.ListenUDP(priv *ecdsa.PrivateKey, conn conn, nodeDBPath string, netrestrict *netutil.Netlist)
+```
+
+We listen on a UDP port for incoming messages from other nodes. This returns a network struct though which we can do this.
+ 
+```go
+network.SetFallbackNodes(BootNode)
+```
+
+This would set our provided bootnode as the fallback node since our in-memory database has no peers stored and therefore none to connect with yet, this populates the local table with all the node records stored in the bootnode. We would need to define a specific Topic for beacon nodes to communicate from, otherwise any peer using Discovery5 protocol would be our peer whether they are participating in Eth2 or not. Topic Advertisement and Search would come up in peer discovery for shards where each node will accept topic registrations from other nodes for specific shards and then relay them to other nodes who are searching for specific shards. This is to provide a scalable and reasonably secure solution for sub-networks which cannot afford to have separate bootstrap nodes for each of those subnetworks.
+
+
+**TODO: Add more detail of how to integrate into Prysm.**
+
+
+**Fork Choice**
+
+Regular sync service will be receiving either blocks or attestations. The block or attestation will get propagated to the block chain service methods `ReceiveBlock` and `ReceiveAttestation`. Those two methods call fork choice. From sync service’s point of view, fork choice is a black box.  In the example code below:
+
+```go
+func (rs *RegularSync) receiveAttestation(msg p2p.Message) error {
+  resp := msg.Data.(*pb.AttestationResponse)
+  att := resp.Attestation
+
+  // Send att to blockchain service to run fork choice
+  If err := rs.chainService.ReceiveAttestation(ctx, att); err != nil {..}
+  // Send att to operation service for inclusion to attestation pool
+  rs.operationService.IncomingAttFeed().send(att)
+
+
+
+
+func (rs *RegularSync) receiveBlock(msg p2p.Message) error {
+  resp := msg.Data.(*pb.BlockResponse)
+  blk := resp.Block
+
+  // Send blk to blockchain service to run state transition and fork choice
+  If err := rs.chainService.ReceiveBlock(ctx, blk); err != nil {..}
```

Comment:
These functions are incomplete. Is this intended?

Reply:
Yes, it's just an example to show that a block or attestation will get sent to the chain service's receive block function and receive attestation function